### ISSUE TITLE
How to deal with multiple inputs?

### ISSUE THREAD

**Issue by ziyinyuan** [dynamicslab/pysindy#226]
I have my experiment data in a table: t, Ag1, x1, y1, z1, Ag2, x2, y2, z2. 
Ag1 and Ag2 are the same constants throughout the time. And different Ag will refer to different xyz. So I have 2 inputs: time and Ag.

I am not sure how exactly to put my data into pysindy and train it.
For model.fit(x,t,u),
1. I am assuming I am including the Ag1 and Ag2 in the "x", is that correct?
2. how would I deal with the "u"?

**Comment by Jacob-Stevens-Haas** [comment]
$u$ is for control inputs - values that vary in time and can affect the evolution of other coordinates, but which do not need equations fit for them.  For your use case, it sounds you should leave `u=None`.

When you say "constants throughout the time", do you mean that Ag1 and Ag2 always take the same value?  Do you have multiple trajectories where they differ only across trajectories?  If they are constants across all time and all trajectories, I believe you can leave them out.  The model will calculate constant terms on its own.

It sounds like you'll want to run model.fit twice: once for Ag1, x1, y1, z1, and once for the second set of data, unless x1, y1, z1 affect x2, y2, z2 or vice versa.

**Comment by ziyinyuan** [comment]
Or let me rephrase my question. If I have an equation x=1-cos(alpha*t) while alpha and t are independent variables. How would I be able to train the data? 

---------------------------------------

dt=0.001
t=np.arange(0,10,dt)
alpha=2
x=1-np.cos(alpha*t)

generalized_library = ps.GeneralizedLibrary(
    libraries=[ps.FourierLibrary(n_frequencies=1)],
    inputs_per_library=np.array([[1,1,2]])
)

model=ps.SINDy(
    differentiation_method=ps.FiniteDifference(order=2),
    optimizer=ps.STLSQ(threshold=0.02),
    feature_names=["x","alpha","t"],
    feature_library=generalized_library
               )
model.fit(x,t=t,u=t)
model.print()

------------------------------------------------------------

It is giving me an error: Each row in inputs_per_library must consist of integers between 0 and the number of total input features - 1.
Also I know that if I only have one input (t), I can use 
def u_fun(t):
    return t
But how do I deal with multiple inputs?

**Comment by Jacob-Stevens-Haas** [comment]
The inputs considered by `inputs_per_library` are (a) all elements of the last axis of `x`, and (b) all elements of the last axis of `u`.

There will be a parametrized library in the future, but for now, you can solve this problem by passing `u=alpha*t`:
```python
import pysindy as ps
import numpy as np

dt=0.01
t=np.arange(0,10,dt)
alpha=2
x=1-np.cos(alpha*t)

generalized_library = ps.GeneralizedLibrary(
libraries=[ps.FourierLibrary(n_frequencies=1)],
inputs_per_library=np.array([[0,1]])
)

model=ps.SINDy(
differentiation_method=ps.FiniteDifference(order=2),
optimizer=ps.STLSQ(threshold=0.02),
feature_names=["x","alpha * t"],
feature_library=generalized_library
)
model.fit(x,t=t,u=alpha * t)
model.print()
```
Gives
```
(x)' = 2.000 sin(1 alpha * t)
```

**Comment by ziyinyuan** [comment]
So this is like a cheat way to using u=alpha*t.
How about x=alpha-np.cos(t)? How would I set the u in this case?

**Comment by Jacob-Stevens-Haas** [comment]
Unfortunately, the ConcatLibrary doesn't accept an `inputs_per_library` parameter, which would allow you to do what you want in a more straightforward manner.

1. You would need to concatenate alpha and t (they would need to be the same length, so something like `u = np.concatenate((alpha * np.ones_like(t)`, t), axis=-1).
1. `f1 = lambda alpha, t: alpha + np.cos(t)`
3. `f1_name = lambda alpha, t: f"{alpha} - cos({t})"`
4. `lib=CustomLibrary(library_functions = [f1], function_names = [f1_name])`
5. Put `CustomLibrary` in a GeneralizedLibrary, with `inputs_per_library = np.array([[0,1,2]])`

**Comment by ziyinyuan** [comment]
If t=np.arange(0,10,0.001), then 'x' and 't' will have a length of 10000. By concatenate alpha and t together with the same length, 'u' would have 20000 rows. But control variable u must have same number of rows as x. So how would model fit this?

**Comment by Jacob-Stevens-Haas** [comment]
Sorry, I made a mistake in the previous post with the library function and `inputs_per_library`.  Here's the correct way to achieve what you want with `CustomLibrary`:
```python
import pysindy as ps
import numpy as np
import matplotlib.pyplot as plt

dt=0.1
t=np.arange(0,10,dt)
t_2d = t.reshape((-1,1))
alpha=2
alpha_2d = alpha * np.ones_like(t_2d)
x=1-np.cos(alpha*t)
x = x.reshape((-1,1))

f1 = lambda alpha, t: np.cos(alpha * t)
f1_name = lambda alpha, t: f"(cos({alpha}*{t}))"
f2 = lambda alpha, t: np.sin(alpha * t)
f2_name = lambda alpha, t: f"(sin({alpha}*{t}))"

lib=ps.CustomLibrary(library_functions = [f1, f2], function_names = [f1_name, f2_name])
generalized_library = ps.GeneralizedLibrary(
libraries=[lib],
inputs_per_library=np.array([[1, 2]])
)

model=ps.SINDy(
    differentiation_method=ps.FiniteDifference(order=2),
    optimizer=ps.STLSQ(threshold=0.02),
    feature_names=["x","alpha", "t"],
    feature_library=generalized_library
)
model.fit(x,t=t,u=np.concatenate((alpha_2d, t_2d), axis=1))
model.print()
```
Prints out results:
```
(x)' = 1.987 (sin(alpha*t))
```
I've used a larger timestep simply for speed in the example, so `x` is length 100.    

## Concatenation explanation
The axes of any array in the current release of pysindy have implicit shape `(n_time, n_coordinates)`.*  So you would want to create `u` to be a shape `(10000, 2)` array.  When pysindy concatenates the control variables, it does it along this same axis, so the combined array would be `[x, alpha, t]` of shape `(10000, 3)`.

*This is made explicit in the current version of `master`, where each array is given  `ax_time`, `ax_coord`, `ax_spatial`, and `ax_sample` properties.

**Comment by ziyinyuan** [comment]
Thank you for clarifying this to me.
How about for function 

----------------------------------------

x=alpha-np.cos(t)

---------------------------------------------------------

I tried the code below, but the plot showed that it doesn't trained properly.

------------------------------------------------------------

f1=lambda alpha,t:alpha-np.cos(t)
f1_name=lambda alpha, t:f"({alpha}-cos({t}))"
f2 = lambda alpha, t: alpha-np.sin(t)
f2_name = lambda alpha, t: f"({alpha}-sin({t}))"

--------------------------------------------------------------

**Comment by ziyinyuan** [comment]
Thank you.
Is it possible to include 2 functions together?

----------------------------------------------------------------------

x1=cos(a*t)
x2=cos(b*t)
x=np.concatenate((x1,x2),axis=0) (combine x1 and x2 in 1 column)

a=np.tile(2,len(x)).reshape((-1,1))
b=np.tile(3,len(x)).reshape((-1,1))
u=np.concatenate((a,b,t),axis=1)

----------------------------------------------------------------------

I can generate the general equation out, but I can't simulate the result.
